homeassistant/components/hunterdouglas_powerview/cover.py

```diff
@@ -192,29 +282,55 @@ async def async_stop_cover(self, **kwargs):
         self._async_update_from_command(await self._shade.stop())
         await self._async_force_refresh_state()
 
+    @callback
+    def _clamp_cover_limit(self, target_hass_position):
+        """Dont allow a cover to go into an impossbile position."""
+        # no override required in base
+        return target_hass_position
+
     async def async_set_cover_position(self, **kwargs):
         """Move the shade to a specific position."""
         if ATTR_POSITION not in kwargs:
             return
-        await self._async_move(kwargs[ATTR_POSITION])
+        await self._async_move(self._clamp_cover_limit(kwargs[ATTR_POSITION]))
+
+    @callback
+    def _set_shade_postion(self, target_hass_position):
+        position_one = hass_position_to_hd(target_hass_position)
+        self.set_position_primary(position_one)
```

Comment:
`await self._shade.move(self._set_shade_postion(target_hass_position))` is wrapped in `self._async_update_from_command(` which is going to override the state set here before it is written to the state machine.  

I think we can drop the `self.set_position_primary` since it is never written to the state machine before it is  changed again when `self._shade.move` returns.

Generally we wait to update the internal state until we know the command we sent to the device was successful.

Maybe this should be called `construct_shade_position` ?